events/renderers/docx.py

```diff
@@ -0,0 +1,269 @@
+            return '%s-%s' % (
+                start.strftime(short_date),
+                end.strftime(short_date),
+            )
+
+        # Multiple years, e.g. 1.1.2017-31.12.2019
+        return '%s-%s' % (
+            start.strftime(long_date),
+            end.strftime(long_date),
+        )
+
+    def __lt__(self, other):
+        return self.start < other.start
+
+    def __eq__(self, other):
+        return self.start == other.start and self.end == other.end
+
+    def __hash__(self):
+        return hash(str(self.start) + str(self.start))
+
+
+def group_by_date(events):
+    # We want events on the same date to be under the same day headline,
+    # and events that span multiple dates to be under their own headlines.
+    events = sorted(events, key=lambda e: e['start_time'])
+
+    dates = {}
+
+    previous_daterange = None
+    for event in events:
+        date_range = DateRange(
+            event['start_time'].date(),
+            event['end_time'].date(),
+            previous=previous_daterange,
+        )
+        previous_daterange = date_range
+
+        if not dates.get(date_range):
+            dates[date_range] = []
+        dates[date_range].append(event)
+
+    dates = OrderedDict(sorted(dates.items()))
+
+    return dates
+
+
+class DOCXRenderer(renderers.BaseRenderer):
+    media_type = (
+        'application/vnd.openxmlformats-officedocument'
+        '.wordprocessingml.document'
+    )
+    format = 'docx'
+    charset = None
+    render_style = 'binary'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        document = Document()
+        parsed_events = []
+
+        query_params = renderer_context['request'].query_params
+
+        event_parser = EventParser()
+
+        if len(data) < 1:
+            raise APIException('No results.')
+
+        first_location = data[0]['location']
+        for raw_event in data:
+            parsed_events.append(event_parser.parse_event(raw_event))
+
+            if raw_event['location'] != first_location:
+                raise APIException('Only one location allowed.')
+
+        # This is here to allow for this to be expanded to include multiple
+        # locations in the future.
+        locations = group_by_location(parsed_events)
+
+        for location in locations:
+            locations[location] = group_by_date(locations[location])
+
+        # We need to get the daterange for the entire document, which is
+        # determined by either the query or the actual events.
+        try:
+            query_start_date = dateutil.parser.parse(query_params.get('start'))
+        except:
+            query_start_date = event_parser.earliest_date
+
+        try:
+            query_end_date = dateutil.parser.parse(query_params.get('end'))
+        except:
+            query_end_date = event_parser.latest_date
+
+        total_date_range = DateRange(query_start_date, query_end_date)
+
+        for location, dateranges in locations.items():
```

Comment:
Places aren't sortable yet and this isn't supposed to handle multiple locations, I just added that because it was easier to profile the code that way (stress testing by bypassing all the location restrictions to get all the events) and because multiple locations might eventually be added. We would also need to figure out things like what the filename would be for multiple locations.